restapi/src/main/java/io/stargate/web/docsapi/service/DocsShredder.java

```diff
@@ -173,10 +174,15 @@ private Object convertToBackendBooleanValue(boolean value, boolean numericBoolea
       String key,
       boolean patching,
       boolean numericBooleans) {
+    // the converter can not handle empty field names, thus add manually
     String bracketedPath = DocsApiUtils.convertJsonToBracketedPath(parsingContext.getJsonPath());
```

Comment:
What happens if there is an empty string that isn't at the leaf-most field name? Such as `{"": {"a": "b"}}`?

It seems that line 180 handles the leaf-most fields but not empty strings that are elsewhere in the document, is that handled properly by `convertJsonToBracketedPath`? (also, I believe there are unit tests for `convertJsonToBracketedPath`, if you could add a test for this)

Reply:
True I guess this is not covered, will check on Monday. :+1:

Reply:
@EricBorczuk I tried fixing the above, however, it's actually impossible with the current approach using surfer.. It fails when there is double empty path and there is nothing we can do about it -> `{"": {"": "value}}`.

The further I looked, I was more confident that the `JsonSurfer` is actually not good fit for document shredding. It does a very bad job of defining the path you are in.. Converting that path to bracketed and then again to array is also very unintuitive. Was there any other idea that was consider here? Why didn't we simply traverse the `JsonNode` here and simply apply handling based on the node type. It would save us a lot of time as well that we spend in checking for different states and conditions?

@tatu-at-datastax any idea if there is a utility for traversing the JSON tree? We need something better than that `JsonSurfer`

Reply:
JsonSurfer was used from the very beginning of the docs API (even before I started on it) - so I don't have a good reason for why it is used. More than likely we could toss it out in favor of a simple traversal of Jackson nodes...and I guess that would allow data like `{"": {"": "value"}}`

But also, to reiterate - is this actually what we want? If we allow storing of empty string (rather than erroring out), will users ever be using this seriously, or will they always be using it by mistake? In other words, what is the use-case for storing empty-string as a field name in JSON (I know it's valid JSON, in practice though it seems like people would only be doing this by mistake)

Reply:
@ivansenic when I looked at code earlier, I had the feeling that JsonSurfer might indeed be sort of overkill; and also add some overhead we could avoid. This because it offers more advanced filtering, traversal, but ones that we either don't need, or can implement more efficiently to fit our exact use case.

I would probably operate directly on JsonNode and not try to find higher level abstractions because I think that our traversal is quite distinct so general-purpose tools would not be good matches.

Reply:
@tatu-at-datastax @EricBorczuk Would you two agree on the following:

1. Do allow storing of empty keys as this is a valid JSON and thus we correctly serve the users?
2. Completely refactor the shredding and implement as Tatu suggested plain simple recursive `JsonNode` traversal? We would even have more dedicated testing possibilities then per node type.

Reply:
On internal replacement for `JsonNode`, I think @ivansenic 's point on keeping track of path would be relevant: our own "StargateJsonValue" (or whatever) could hold on to path indicator(s) as needed; avoid need for String concatenation and so forth.

Reply:
Agree on (2) as that could have multiple benefits, including simpler implementation.

I am bit ambivalent on (1) -- use of empty String as key is valid JSON and if we are confident we can make it work 100%, I'd be for it. My main concern is whether there are cases where we simply cannot make it work reliably, or that are there are ambiguous cases. I think you @ivansenic and @EricBorczuk are in better position to decide on this one.

Reply:
The main issue I see with (1) is that there's no difference between a missing pN column value and one that is set to empty string because a user supplied it (i.e. we don't store `null` for missing clustering columns because that isn't allowed) - so in the next step when we want wild-card `*` matching to work with empty string fields (which internally adds a clause `WHERE pN > ''`), we won't really be able to get empty-string fields to match.

Reply:
What about storing a fixed value, like we do for the root document and empty arrays/maps?

Reply:
Do we store actual String values as-is, or JSON-escaped ones, in leaf value fields? If former, it may be difficult to find any placeholder value that could not be stored by users.

Reply:
@tatu-at-datastax Yea we have those already, have a look here https://github.com/stargate/stargate/blob/7bfee16e3767646541e3cc3f982fc5775e1d2194/restapi/src/main/java/io/stargate/web/docsapi/service/query/DocsApiConstants.java#L46-L52